massa-execution-worker/src/speculative_async_pool.rs

```diff
@@ -351,6 +342,114 @@ impl SpeculativeAsyncPool {
         msgs
     }
 
+    /// In this version, we need to ensure the order of the messages is preserved when we fetch them
+    /// (to avoid a non-deterministic behavior in the execution).
+    fn fetch_msgs_v1(
+        &mut self,
+        wanted_ids: Vec<&AsyncMessageId>,
+        delete_existing: bool,
+        execution_component_version: u32,
+    ) -> Vec<(AsyncMessageId, Option<AsyncMessage>)> {
+        let mut msgs_opt = Vec::new();
+        let mut still_wanted_ids_index = Vec::new();
+
+        let mut current_changes = HashMap::new();
+        for id in wanted_ids.iter() {
+            current_changes.insert(*id, AsyncMessageUpdate::default());
+        }
+
+        let pool_changes_clone = self.pool_changes.clone();
+
+        // First, look in speculative pool
+        for (index, message_id) in wanted_ids.iter().enumerate() {
+            match pool_changes_clone.0.get(message_id) {
+                Some(SetUpdateOrDelete::Set(msg)) => {
+                    if delete_existing {
+                        self.pool_changes.push_delete(**message_id);
+                    }
+                    msgs_opt.push((**message_id, Some(msg.clone())));
+                }
+                Some(SetUpdateOrDelete::Update(msg_update)) => {
+                    current_changes.entry(message_id).and_modify(|e| {
+                        e.apply(msg_update.clone());
+                    });
+                    msgs_opt.push((**message_id, None));
+                    still_wanted_ids_index.push(index);
+                }
+                Some(SetUpdateOrDelete::Delete) => {
```

Comment:
Delete and None has same code 
```suggestion
              Some(SetUpdateOrDelete::Delete) | None => {
```